core/store/src/flat_state.rs

```diff
@@ -352,49 +357,130 @@ pub struct FlatStorageState(Arc<RwLock<FlatStorageStateInner>>);
     #[allow(unused)]
     shard_id: ShardId,
-    /// The block for which we store the key value pairs of its state. Note that it is not necessarily
-    /// the latest block.
+    /// The block for which we store the key value pairs of its state. For non catchup mode,
+    /// it should be the last final block.
     #[allow(unused)]
-    head: CryptoHash,
-    /// This is the oldest block that whose state is supported by flat storage. In normal block
-    /// processing mode, this should be the latest final block.
+    flat_head: CryptoHash,
+    /// Stores some information for all blocks supported by flat storage, this is used for finding
+    /// paths between the root block and a target block
     #[allow(unused)]
-    tail: CryptoHash,
-    /// Stores a copy for all block headers for blocks that flat storage supports. This is useful
-    /// for finding the path between the current head to a target block. It's possible that we don't
-    /// need all the information in block header and this can be simplified later, but for now,
-    /// for simplicity we store the entire BlockHeader.
-    #[allow(unused)]
-    block_headers: HashMap<CryptoHash, BlockHeader>,
+    blocks: HashMap<CryptoHash, BlockInfo>,
     /// State deltas for all blocks supported by this flat storage.
     /// All these deltas here are stored on disk too.
     #[allow(unused)]
     deltas: HashMap<CryptoHash, FlatStateDelta>,
 }
 
+#[cfg(feature = "protocol_feature_flat_state")]
+pub mod store_helper {
+    use crate::flat_state::KeyForFlatStateDelta;
+    use crate::{FlatStateDelta, Store, StoreUpdate};
+    use borsh::BorshSerialize;
+    use near_primitives::hash::CryptoHash;
+    use near_primitives::types::ShardId;
+
+    pub(crate) fn get_delta(
+        store: &Store,
+        shard_id: ShardId,
+        block_hash: CryptoHash,
+    ) -> Result<Option<FlatStateDelta>, crate::StorageError> {
+        let key = KeyForFlatStateDelta { shard_id, block_hash };
+        store
+            .get_ser(crate::DBCol::FlatStateDeltas, &key.try_to_vec().unwrap())
+            .map_err(|_| crate::StorageError::StorageInternalError)
+    }
+
+    pub fn set_delta(
+        store_update: &mut StoreUpdate,
+        shard_id: ShardId,
+        block_hash: CryptoHash,
+        delta: &FlatStateDelta,
+    ) -> Result<(), crate::StorageError> {
+        let key = KeyForFlatStateDelta { shard_id, block_hash };
+        store_update
+            .set_ser(crate::DBCol::FlatStateDeltas, &key.try_to_vec().unwrap(), delta)
+            .map_err(|_| crate::StorageError::StorageInternalError)
+    }
+
+    pub(crate) fn get_flat_head(store: &Store, shard_id: ShardId) -> CryptoHash {
+        store
+            .get_ser(crate::DBCol::FlatStateMisc, &shard_id.try_to_vec().unwrap())
+            .expect("Error reading flat head from storage")
+            .expect(&format!("Cannot read flat head for shard {} from storage", shard_id))
```

Comment:
maybe avoid allocation on success path?
```suggestion
            .unwrap_or_else(|| panic!("Cannot read flat head for shard {} from storage", shard_id))
```